File: Tv2 key prep .py
"""
tv2_key_prep.py
Thành viên 2 - Cấu trúc toán học & Tiền xử lý khóa HMAC
--------------------------------------------------------------
Nhiệm vụ:
  1. Chuẩn hóa khóa K  →  K'  (băm nếu quá dài, đệm 0x00 nếu quá ngắn)
  2. Tính Inner Key  =  K' ⊕ ipad  (0x36 lặp block_size lần)
  3. Tính Outer Key  =  K' ⊕ opad  (0x5C lặp block_size lần)
  4. In tất cả kết quả dưới dạng Hex
"""

import hashlib

#Hằng số

IPAD_BYTE = 0x36
OPAD_BYTE = 0x5C

#Bảng thông số cho từng hàm băm được hỗ trợ
HASH_PARAMS = {
    "sha256": {"block_size": 64,  "output_size": 32,  "fn": hashlib.sha256},
    "sha512": {"block_size": 128, "output_size": 64,  "fn": hashlib.sha512},
    "sha1":   {"block_size": 64,  "output_size": 20,  "fn": hashlib.sha1},
    "md5":    {"block_size": 64,  "output_size": 16,  "fn": hashlib.md5},
}

#Hàm tiện ích

def to_bytes(data) -> bytes:
    """Chuyển string hoặc bytes sang bytes."""
    if isinstance(data, str):
        return data.encode("utf-8")
    return bytes(data)


def hex_dump(label: str, data: bytes, bytes_per_line: int = 16) -> None:
    """In dữ liệu dạng hex có nhãn, mỗi dòng bytes_per_line byte."""
    hex_str = data.hex()
#Nhóm thành từng cặp(1 byte)
    pairs = [hex_str[i:i+2] for i in range(0, len(hex_str), 2)]
    lines = [" ".join(pairs[i:i+bytes_per_line])
             for i in range(0, len(pairs), bytes_per_line)]
    print(f"\n{'─'*60}")
    print(f"  {label}  [{len(data)} bytes]")
    print(f"{'─'*60}")
    for line in lines:
        print(f"  {line}")

#Hàm 1 : Chuẩn hóa khóa K  → K'

def prepare_key(key: bytes, block_size: int, hash_fn) -> bytes:
    """
    Chuẩn hóa khóa K thành K' có độ dài đúng bằng block_size.

    Quy tắc (RFC 2104):
      - Nếu len(K) > block_size  →  K' = H(K)        (băm để thu gọn)
      - Nếu len(K) <= block_size →  K' = K            (giữ nguyên)
      - Sau đó đệm 0x00 bên phải cho đến khi len(K') = block_size
    """
    if len(key) > block_size:
        print(f"\n  [!] Khóa dài hơn block_size ({len(key)} > {block_size})")
        print(f"  [!] Tiến hành băm khóa K trước...")
        key = hash_fn(key).digest()
        print(f"  [!] Sau khi băm: len(K) = {len(key)} bytes")

#Đệm 0x00 bên phải
    key_prime = key.ljust(block_size, b"\x00")
    return key_prime

#Hàm 2 : Tính Inner Key = K' ⊕ ipad

def compute_inner_key(key_prime: bytes, block_size: int) -> bytes:
    """
    Tính inner_key = K' XOR ipad.
    ipad = byte 0x36 lặp lại block_size lần.
    """
    ipad = bytes([IPAD_BYTE] * block_size)
    inner_key = bytes(k ^ i for k, i in zip(key_prime, ipad))
    return inner_key

#Hàm 3 : Tính Outer Key = K' ⊕ opad

def compute_outer_key(key_prime: bytes, block_size: int) -> bytes:
    """
    Tính outer_key = K' XOR opad.
    opad = byte 0x5C lặp lại block_size lần.
    """
    opad = bytes([OPAD_BYTE] * block_size)
    outer_key = bytes(k ^ o for k, o in zip(key_prime, opad))
    return outer_key

#Hàm tổng hợp : chạy toàn bộ bước tiền xử lý và in kết quả

def key_prep_demo(key, hash_name: str = "sha256"):
    """
    Chạy toàn bộ bước tiền xử lý khóa của HMAC và in ra kết quả.

    Tham số:
        key       : str hoặc bytes – khóa bí mật K
        hash_name : str – tên hàm băm ("sha256", "sha512", "sha1", "md5")
    """
    if hash_name not in HASH_PARAMS:
        raise ValueError(f"Hàm băm '{hash_name}' chưa được hỗ trợ. "
                         f"Chọn một trong: {list(HASH_PARAMS.keys())}")

    params     = HASH_PARAMS[hash_name]
    block_size = params["block_size"]
    hash_fn    = params["fn"]
    key_bytes  = to_bytes(key)

    print("=" * 60)
    print(f"  HMAC KEY PREPARATION  –  Hàm băm: {hash_name.upper()}")
    print("=" * 60)
    print(f"  Block size  : {block_size} bytes ({block_size * 8} bits)")
    print(f"  Output size : {params['output_size']} bytes ({params['output_size'] * 8} bits)")
    print(f"  ipad        : 0x{IPAD_BYTE:02X}  (lặp {block_size} lần)")
    print(f"  opad        : 0x{OPAD_BYTE:02X}  (lặp {block_size} lần)")

#Khóa gốc K
    hex_dump("K (khóa gốc)", key_bytes)

#Bước 1 : Chuẩn hóa K → K'
    print(f"\n>>> BƯỚC 1: Chuẩn hóa K → K'")
    key_prime = prepare_key(key_bytes, block_size, hash_fn)
    hex_dump("K' (sau khi đệm đến block_size)", key_prime)

#Bước 2 : Inner Key
    print(f"\n>>> BƯỚC 2: Inner Key  =  K' ⊕ ipad (0x{IPAD_BYTE:02X})")
    inner_key = compute_inner_key(key_prime, block_size)
    hex_dump("Inner Key  (K' ⊕ ipad)", inner_key)

#Bước 3 : Outer Key
    print(f"\n>>> BƯỚC 3: Outer Key  =  K' ⊕ opad (0x{OPAD_BYTE:02X})")
    outer_key = compute_outer_key(key_prime, block_size)
    hex_dump("Outer Key  (K' ⊕ opad)", outer_key)

    print("\n" + "=" * 60)
    print("  Hoàn tất tiền xử lý khóa.")
    print("  Inner Key và Outer Key sẵn sàng cho bước tính HMAC.")
    print("=" * 60)

    return key_prime, inner_key, outer_key

#Demo với các test case từ báo cáo

if __name__ == "__main__":

    print("\n" + "█" * 60)
    print("  TEST CASE 1 – Khóa ngắn: 'key'  |  SHA-256")
    print("█" * 60)
    key_prep_demo("key", "sha256")

    print("\n" + "█" * 60)
    print("  TEST CASE 2 – Khóa rỗng: ''  |  SHA-256")
    print("█" * 60)
    key_prep_demo("", "sha256")

    print("\n" + "█" * 60)
    print("  TEST CASE 3 – Khóa quá dài (100 bytes)  |  SHA-256")
    print("█" * 60)
    long_key = "A" * 100      # 100 bytes > block_size=64 → phải băm
    key_prep_demo(long_key, "sha256")

    print("\n" + "█" * 60)
    print("  TEST CASE 4 – Khóa 'secret'  |  SHA-256")
    print("█" * 60)
    key_prep_demo("secret", "sha256")

File: test/Tv2 key prep .cpp
/*
 * tv2_key_prep.cpp
 * Thành viên 2 – Cấu trúc toán học & Tiền xử lý khóa HMAC
 * ─────────────────────────────────────────────────────────
 * Biên dịch:
 *   g++ -std=c++17 tv2_key_prep.cpp -lssl -lcrypto -o tv2_key_prep
 * Chạy:
 *   ./tv2_key_prep
 * ─────────────────────────────────────────────────────────
 * Nhiệm vụ:
 *   1. Chuẩn hóa khóa K  →  K'
 *        - Băm K nếu len(K) > block_size  (RFC 2104)
 *        - Đệm 0x00 bên phải đến đúng block_size
 *   2. Tính Inner Key  =  K' ⊕ ipad  (0x36 × block_size)
 *   3. Tính Outer Key  =  K' ⊕ opad  (0x5C × block_size)
 *   4. In tất cả giá trị dưới dạng Hex
 */

#include <iostream>
#include <iomanip>
#include <vector>
#include <string>
#include <stdexcept>
#include <openssl/evp.h>   // EVP_MD_CTX, EVP_sha256, …
#include <openssl/sha.h>


//   HẰNG SỐ & KIỂU DỮ LIỆU

static constexpr uint8_t IPAD_BYTE = 0x36;
static constexpr uint8_t OPAD_BYTE = 0x5C;

using Bytes = std::vector<uint8_t>;

// Thông số cho từng hàm băm 
struct HashParams {
    std::string  name;
    size_t       block_size;   // bytes
    size_t       output_size;  // bytes
    const EVP_MD* (*evp_fn)(); // con trỏ hàm trả về EVP_MD*
};

static const HashParams HASH_TABLE[] = {
    { "SHA-256", 64,  32, EVP_sha256 },
    { "SHA-512", 128, 64, EVP_sha512 },
    { "SHA-1",   64,  20, EVP_sha1   },
    { "MD5",     64,  16, EVP_md5    },
};


//   HÀM TIỆN ÍCH


//Chuyển std::string sang Bytes 
Bytes to_bytes(const std::string& s) {
    return Bytes(s.begin(), s.end());
}

// In dữ liệu dạng hex, bytes_per_line byte mỗi dòng 
void hex_dump(const std::string& label, const Bytes& data,
              size_t bytes_per_line = 16)
{
    std::string sep(60, '-');
    std::cout << "\n" << sep << "\n";
    std::cout << "  " << label << "  [" << data.size() << " bytes]\n";
    std::cout << sep << "\n";

    for (size_t i = 0; i < data.size(); ++i) {
        if (i > 0 && i % bytes_per_line == 0)
            std::cout << "\n";
        std::cout << "  " << std::hex << std::setw(2)
                  << std::setfill('0') << static_cast<int>(data[i]);
    }
    if (!data.empty()) std::cout << "\n";
    std::cout << std::dec; // reset về decimal
}


//   HÀM BĂNG (wrapper dùng OpenSSL EVP)

Bytes hash_data(const Bytes& data, const EVP_MD* md) {
    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    if (!ctx) throw std::runtime_error("EVP_MD_CTX_new() thất bại");

    unsigned int len = 0;
    Bytes digest(EVP_MD_size(md));

    if (EVP_DigestInit_ex(ctx, md, nullptr) != 1 ||
        EVP_DigestUpdate(ctx, data.data(), data.size()) != 1 ||
        EVP_DigestFinal_ex(ctx, digest.data(), &len) != 1)
    {
        EVP_MD_CTX_free(ctx);
        throw std::runtime_error("Hàm băm thất bại");
    }
    EVP_MD_CTX_free(ctx);
    digest.resize(len);
    return digest;
}


//   HÀM 1: Chuẩn hóa khóa K → K'  (RFC 2104)

/*
 * Quy tắc:
 *   - Nếu len(K) > block_size  →  K' = H(K)   (băm để thu gọn)
 *   - Nếu len(K) <= block_size →  K' = K
 *   - Đệm 0x00 bên phải cho đến len(K') = block_size
 */
Bytes prepare_key(const Bytes& key, const HashParams& params) {
    Bytes k = key;

    if (k.size() > params.block_size) {
        std::cout << "\n  [!] Khóa dài hơn block_size ("
                  << k.size() << " > " << params.block_size << ")\n";
        std::cout << "  [!] Tiến hành băm khóa K trước...\n";
        k = hash_data(k, params.evp_fn());
        std::cout << "  [!] Sau khi băm: len(K) = " << k.size() << " bytes\n";
    }

    // Đệm 0x00 bên phải
    k.resize(params.block_size, 0x00);
    return k;
}


//   HÀM 2: Inner Key  =  K' ⊕ ipad

Bytes compute_inner_key(const Bytes& key_prime, size_t block_size) {
    Bytes inner(block_size);
    for (size_t i = 0; i < block_size; ++i)
        inner[i] = key_prime[i] ^ IPAD_BYTE;
    return inner;
}


//   HÀM 3: Outer Key  =  K' ⊕ opad

Bytes compute_outer_key(const Bytes& key_prime, size_t block_size) {
    Bytes outer(block_size);
    for (size_t i = 0; i < block_size; ++i)
        outer[i] = key_prime[i] ^ OPAD_BYTE;
    return outer;
}


//   HÀM TỔNG HỢP: chạy toàn bộ và in kết quả

void key_prep_demo(const std::string& key_str,
                   const HashParams&  params)
{
    std::string sep(60, '=');
    std::cout << "\n" << sep << "\n";
    std::cout << "  HMAC KEY PREPARATION  –  Hàm băm: " << params.name << "\n";
    std::cout << sep << "\n";
    std::cout << "  Block size  : " << params.block_size
              << " bytes (" << params.block_size * 8 << " bits)\n";
    std::cout << "  Output size : " << params.output_size
              << " bytes (" << params.output_size * 8 << " bits)\n";
    std::cout << "  ipad        : 0x"
              << std::hex << std::setw(2) << std::setfill('0')
              << static_cast<int>(IPAD_BYTE) << std::dec
              << "  (lặp " << params.block_size << " lần)\n";
    std::cout << "  opad        : 0x"
              << std::hex << std::setw(2) << std::setfill('0')
              << static_cast<int>(OPAD_BYTE) << std::dec
              << "  (lặp " << params.block_size << " lần)\n";

    Bytes key = to_bytes(key_str);

    //  Khóa gốc K 
    hex_dump("K (khóa gốc)", key);

    // Bước 1: K → K' 
    std::cout << "\n>>> BƯỚC 1: Chuẩn hóa K → K'\n";
    Bytes key_prime = prepare_key(key, params);
    hex_dump("K' (sau khi đệm đến block_size)", key_prime);

    // Bước 2: Inner Key
    std::cout << "\n>>> BƯỚC 2: Inner Key  =  K' ⊕ ipad (0x"
              << std::hex << std::setw(2) << std::setfill('0')
              << static_cast<int>(IPAD_BYTE) << std::dec << ")\n";
    Bytes inner_key = compute_inner_key(key_prime, params.block_size);
    hex_dump("Inner Key  (K' ⊕ ipad)", inner_key);

    //  Bước 3: Outer Key 
    std::cout << "\n>>> BƯỚC 3: Outer Key  =  K' ⊕ opad (0x"
              << std::hex << std::setw(2) << std::setfill('0')
              << static_cast<int>(OPAD_BYTE) << std::dec << ")\n";
    Bytes outer_key = compute_outer_key(key_prime, params.block_size);
    hex_dump("Outer Key  (K' ⊕ opad)", outer_key);

    std::cout << "\n" << sep << "\n";
    std::cout << "  Hoàn tất tiền xử lý khóa.\n";
    std::cout << "  Inner Key và Outer Key sẵn sàng cho bước tính HMAC.\n";
    std::cout << sep << "\n";
}


//   MAIN – CHẠY 4 TEST CASE THEO BÁO CÁO

int main() {
    const HashParams& sha256 = HASH_TABLE[0]; // SHA-256

    auto banner = [](const std::string& msg) {
        std::string line(60, 0x23); // '#'
        std::cout << "\n" << line << "\n  " << msg << "\n" << line << "\n";
    };

    //  Test 1 
    banner("TEST CASE 1 - Khoa ngan: \"key\"  |  SHA-256");
    key_prep_demo("key", sha256);

    //  Test 2 
    banner("TEST CASE 2 - Khoa rong: \"\"  |  SHA-256");
    key_prep_demo("", sha256);

    // Test 3 – khóa 100 bytes > block_size=64 
    banner("TEST CASE 3 - Khoa qua dai (100 bytes)  |  SHA-256");
    key_prep_demo(std::string(100, 'A'), sha256);

    //  Test 4 
    banner("TEST CASE 4 - Khoa \"secret\"  |  SHA-256");
    key_prep_demo("secret", sha256);

    return 0;
}
